### ISSUE TITLE
[Bug]: signal never gets aborted upon navigation for non-awaited deferred loader

### ISSUE THREAD

**Issue by mz8i** [remix-run/react-router#10252]
### What version of React Router are you using?

6.9.0

### Steps to Reproduce

```ts

function fetchA(signal: AbortSignal) {
    signal.addEventListener('abort', () => console.log('Fetcher A cancelled'));

    return Promise.resolve('a');
}

function fetchB(signal: AbortSignal) {
    signal.addEventListener('abort', () => console.log('Fetcher B cancelled'));

    return new Promise(() => {});
}

export function loader({request: {signal}}: LoaderFunctionArgs) {
    return defer({
        a: await fetchA(signal),
        b: fetchB(signal)
    });
}
```

### Expected Behavior

In the example deferred loader setup above, with some awaited load functions and other non-awaited functions that take a potentially long time to resolve, the abort signal passed to `fetchB` should be aborted when navigating away from the page before the loader completes.

### Actual Behavior

If the awaited functions have resolved, the signal passed to `fetchB` is not aborted upon navigation.

Having a brief look at the code, I think this happens because on line https://github.com/remix-run/react-router/blob/5c03b9c540340d861d50895b443b37fda0a744d0/packages/router/router.ts#L1278 the `pendingNavigationController` is cleared with the assumption that the loaders have completed, which isn't necessarily true in the deferred case. Because of that, when the new navigation starts, the signal is never aborted https://github.com/remix-run/react-router/blob/5c03b9c540340d861d50895b443b37fda0a744d0/packages/router/router.ts#L1165

It seems to me that to properly support the deferred use case, `pendingNavigationController` should simply not be cleared - the signal would then be aborted upon every navigation, but in most cases there would be no loader functions listening to it any more, so nothing would happen.

**Comment by nakah** [comment]
I'm currently facing this issue with defer & signal not aborted.
Is there a workaround we can implement at least ?

**Comment by hellyet** [comment]
Same here.

**Comment by mz8i** [comment]
@ryanflorence sorry for tagging directly, but is there a chance this could at least be triaged? It doesn't seem there's been any activity from the team here since this was posted more than a year ago. Happy to provide more context or a fully reproducible example if needed.

I consider this a fairly serious issue, because if there are expensive loading operations inside the `defer` call, they won't get cancelled upon navigating away from the current view.

In an example case where each page starts to load a large dataset, but the user might quickly navigate away to a different page, this will lead to wasting the user's bandwidth on a lot of unnecessary data loading, potentially leading to a degraded experience...

If my diagnosis of the cause of the problem is correct, then the solution suggested in the original post still seems to apply, even though the exact line numbers have changed since.

**Comment by jcheese1** [comment]
maybe @brophdawg11 ?

**Comment by jcheese1** [comment]
bumping this

**Comment by brophdawg11** [comment]
I don't consider this a bug - the current logic considers the resolution of the route loaders the "end" of the navigation.  The destination location now has all required data, is rendered, and the URL can be updated.  Navigating away _after_ that is not an interruption of the prior navigation - it's just a new navigation.

Here's a relevant comment from @jacob-ebey  in Discord ([link](https://discord.com/channels/770287896669978684/909977478133583963/1335423567374061568)) in response to this type of request:

> TLDR; Aborting defered data after it has surfaced leads to errors surfacing, and therefore can't be aborted.
> You click a link and the loader resolves with nothing but promises. You render those promises in <Await> and the user clicks a link in a parent route that aborts the navigation. The promise passed to <Await> immidiately rejects and surfaces the error.

But even if we ignore that for a moment, this is not as simple as it might seem. Your proposed solution might work for non-nested routing scenarios (in theory).  But to achieve the same result in a nested routes setup means route-specific signals, route specific requests, and hanging onto potentially many prior navigation controllers well beyond the completion of their navigations.

Consider a simple nested routing scenario of a parent route with 2 children (`/parent/a` and `/parent/b`).  Assume all routes return non-awaited streaming promises as above.

1. When the user navigates from `/` to `/parent/a` and we land on `/a`, we have outstanding promises on both the `parent` `loader` and the `a` `loader`
2. Before those resolve, the user then navigates to `/parent/b`
3. This will not re-call the `parent` `loader` because the route is reused, but it will call the `b` `loader`
4. So, if we abort the `signal` from (1), that will abort the pending promises for both `parent` and `a`
5. Uh oh!  Because parent's `loader` isn't re-called, the non-awaited promise we just aborted will _never_ resolve and the related `Suspense` boundary around `<Await>` in the parent will be stuck in the fallback state forever

Even re-calling the parent `loader` is not a proper solution because you lose any progress you made prior to navigating to `/parent/b` and you cal the loader all over again and start from scratch - unnecessarily delaying the UI you wanted to show to the user in the parent layout.

This only gets more complex with multiple levels of nesting, having to track many prior controllers/signals for many prior navigations, tracking which routes first appeared on which navigation, and if they have since revalidated, so we could potentially abort the _original_ signals at the right times.
